src/internal/m365/collection/drive/permission.go

```diff
@@ -509,5 +519,9 @@ func RestorePermissions(
 		return nil
 	}
 
-	return clues.Wrap(err, "updating permissions").OrNil()
+	if err != nil {
+		errs.AddRecoverable(ctx, clues.WrapWC(ctx, err, "updating permissions"))
+	}
+
+	return nil
```

Comment:
Some concerns about the pattern of producing both recoverable errors and also returning an error here.  This is a usage we've tried to stay away from in the past because it makes error ownership ambiguous.

The base idea is recoverable errors allow the owner of a loop to complete the loop without exiting early from an error.  Non-looping code should simply return errors (as is idiomatic in golang) because that scope doesn't have full context about whether the caller considers that call recoverable.  Now that recoverable errors are expected to produce top-level operation failures (depending on configs), it can be risky for arbitrary funcs to assume identification and ownership of recoverability.

In this case, I think we should retain the original error handling instead of adding recoverables.  If the case is something we can move past without exiting the func over (ie, line 488), then an info log provides sufficient tracking.

Reply:
Hmm, I understand your concern but this one is a little tricky. If we fail to create the folders, we can't proceed(and so we return an error), but we need both link share and permissions restore to go through even when there are errors. Also, we should not consider the restore a success with a link share or permissions restore error and so just using a log would not work here.

Here is what we want the code to do. Let me know if you have a suggestion on how to better do this:
- If folder creation fails, hard error
- If link share or permissions restore fails, continue but with recoverable error

Reply:
Ahh, you are right. There is no reason why `RestorePermissions` should be returning an error.

Reply:
Hmm, the problem is that within `RestorePermissins`, I want to ensure we restore permissions even if restore of link shares fail. If we return an error, this is not possible. One option would be to capture and return the link share error at the end if permissions succeed and and link shares fail, but this might look non-idiomatic as well.

Reply:
> Hmm, the problem is that within RestorePermissins, I want to ensure we restore permissions even if restore of link shares fail. If we return an error, this is not possible

In that case it shouldn't produce an error.  Only a log.  Maybe an alert?  Recoverable errors should be thought of as eqivalent to non-recoverable errors.  The only difference is that we might defer the handling, depending on runtime configuration.  IE: if we wouldn't exit the process with an error, we don't need a recoverable error.

Reply:
> ... if we wouldn't exit the process with an error, we don't need a recoverable error.

We have to exit the process with an error, but only after we have completed restoring everything we can.